File: ruby/lib/pay_kit/fee.rb
# frozen_string_literal: true

require_relative "errors"
require_relative "price"

module PayKit
  # Single fee entry: a recipient address and what they receive.
  Fee = Data.define(:recipient, :price, :kind) do
    # Whether this fee is taken out of the gate's amount (`:within`)
    # or added on top of it (`:on_top`).
    def within?
      kind == :within
    end

    def on_top?
      kind == :on_top
    end
  end

  # Build Fee arrays from the `{ recipient => Price }` hash kwargs
  # accepted by `gate(...)`. Coerces user input and validates the
  # shape before the Gate sees it.
  module FeeBuilder
    module_function

    def from_hash(hash, kind:)
      return [].freeze if hash.nil?

      unless hash.is_a?(Hash)
        raise ConfigurationError, "fee_#{kind} must be a Hash of { recipient => price }"
      end

      hash.map do |recipient, price|
        unless recipient.is_a?(String) && !recipient.empty?
          raise ConfigurationError, "fee_#{kind} recipient must be a non-empty String, got #{recipient.inspect}"
        end
        unless price.is_a?(Price)
          raise ConfigurationError, "fee_#{kind} price for #{recipient.inspect} must be built via usd(...)/eur(...)"
        end

        Fee.new(recipient: recipient, price: price, kind: kind)
      end.freeze
    end
  end
end

File: python/src/pay_kit/fee.py
"""A single recipient line on a Gate.

Either taken ``within`` the gate amount (the payTo recipient nets less) or
``on_top`` of it (the customer pays more, payTo nets the full amount). Frozen
at construction; Gate builds these from its ``fee_within`` / ``fee_on_top``
arguments.
"""

from __future__ import annotations

from typing import Literal

import pydantic

from pay_kit.errors import ConfigurationError
from pay_kit.price import Price

__all__ = ["Fee"]


class Fee(pydantic.BaseModel):
    """A recipient address and the price they receive, within or on top."""

    model_config = pydantic.ConfigDict(frozen=True, extra="forbid")

    recipient: str
    price: Price
    kind: Literal["within", "on_top"]

    @pydantic.field_validator("recipient")
    @classmethod
    def _non_empty_recipient(cls, value: str) -> str:
        if not value:
            raise ConfigurationError("pay_kit: Fee recipient must be a non-empty string")
        return value

    def is_within(self) -> bool:
        """Whether this fee is taken out of the gate's amount."""
        return self.kind == "within"

    def is_on_top(self) -> bool:
        """Whether this fee is added on top of the gate's amount."""
        return self.kind == "on_top"
